### PR TITLE
Make zippered for loops more like parallel zippered loops by having the first expression lead

### LINKED ISSUE #20041: `--baseline`/`--no-inline-iterators` sends method iterators on unbounded ranges into an infinite loop
### Summary of Problem

When calling an iterator on an unbounded range with `--no-inline-iterators`, which is a part of `--baseline`, the compiler sends the generated code into an infinite loop because it assumes that any iterator over an unbounded range is itself infinite.  This is no longer a safe assuption now that we're having unbounded ranges over bools and enums simply span the values in question.  But it wasn't even before that since someone could define an obviously finite iterator on such ranges like:

```chapel
iter range.foo() {
  yield 1;
}
```

### Steps to Reproduce

**Associated Future Test(s):**
[`test/functions/iterators/bugs/no-inline-method-iter.chpl`](
           https://github.com/chapel-lang/chapel/blob/main/test/functions/iterators/bugs/no-inline-method-iter.chpl
          ) #20038 

### Configuration Information

- Output of `chpl --version`: `chpl version 1.27.0 pre-release (446195541f)`

### LINKED ISSUE #20051: Do we have the wrong behavior currently for loops over `zip(1.., expr)`? (zipped unbounded ranges)
While writing some code last week, I was surprised to find that a loop like:

```chapel
for (i,j) in zip(1.., 1..5) do
  writeln((i,j));
```

currently loops five times then stops without complaint.  My expectation was that it would loop five times and then complain of a size mismatch between the two iterands.  Specifically, Chapel generally complains about size mismatches in serial loops (and tries to in parallel loops, though it's unsuccessful in some cases, noted in issue #11428); and it seems that it ought to be complaining in this case as well.

In the parallel zippered case, we've traditionally given unbounded ranges in follower positions special "can conform to anything" semantics, which permit loops like:

```chapel
forall (a, i) in zip(my1dArray, 0..) do ...
```

and we've discussed extending leader-follower iterators to permit users to write their own "conforming" iterators, though that has not yet been implemented.  For that reason, it does not surprise me that:

```chapel
for (i,j) in zip(1..5, 1..) do
```

would run five times then terminate without complaint, but I would expect that if an unbounded range is in the first/leader position of the `zip()` that the loop would be conceptually infinite.  Note that this can be useful in cases where the loop contains a break/return/exit, as in this (toy) example:

```chapel
for (i,j) in zip(0.., genRandNums()) {
  if j == 0.5 {
    writeln("Got 0.5 on iteration ", i);
    break;
  }
}
```

This issue asks whether we've made a mistake here, or whether there's some rationale for the current behavior that I'm missing.

### PR DESCRIPTION
**Pull request by bradcray**
This implements the semantics proposed in #20051 in which the
initial / leader iterand in a serial zippered for loop controls the
expectations for the loop, such as how many iterations it runs.
As a particular example, if an unbounded int range is leading, the
loop will run a conceptually unbounded number of iterations, unless
it contains a break/return/halt.  If follower iterands run out of iterations
before the loop terminates, that will also be an error, as in typical
zippered size mismatches.

I'm considering the current behavior to be more of a bug than a
language change (though we've clearly written test codes that
rely on the current behavior), so rather than providing a backwards-
compatible implementation for the old behavior, this PR prints a
warning about the behavior change when an unbounded range
is given as the leader mentioning the change in behavior
and suggesting that the iterands be reordered to preserve the
old behavior.  Since the update is fairly simple and avoided the need
to have two ways of rewriting zippered iterators, this seemed very
reasonable.

Resolves #20051 
Resolves #20041

### CODE REVIEW

File: compiler/resolution/lowerIterators.cpp [hunk 1]
```diff
@@ -2523,27 +2523,32 @@ expandForLoop(ForLoop* forLoop) {
...
-            forLoop->insertAtTail(buildIteratorCall(cond, HASMORE, iterators.v[i], children));
-
-            testBlock = new BlockStmt(new SymExpr(cond));
-          }
+      // If we haven't yet generated a test to terminate the loop and
+      // are in a zippered context, we'll set it up based on this
+      // iterator which is presumably the first.
+      if (testBlock == NULL) {
+        if (!isBoundedIterator(iterFn) && iterators.n > 1) {
+          USR_WARN(forLoop, "The behavior of zippered serial loops whose leader is an unbounded range has changed in this release; to maintain the previous behavior, swap a bounded iterand into the leader position (the first expression in the 'zip(...)').");
```
Comment (vasslitvinov) [inline compiler/resolution/lowerIterators.cpp:2531]: How about we point them to #20051 ? The warning is lacking an explanation of what the change is.

File: test/release/examples/primers/iterators.chpl [hunk 2]
```diff
@@ -52,30 +52,37 @@ writeln("...");
...
+// A zippered loop can have a single index variable, which will be a tuple
+// of the yielded values; or a tuple of variables like ``(i, j)``, each of
+// which is initialized with the value yielded by the corresponding iterator.
+//
+// In zippered loops, the first iterand controls the behavior of the loop,
+// and is referred to as the "leader".  Subsequent iterands must
+// each either match the number of iterations or else be unbounded, as in
+// this example.  When iterating over multidimensional domains and arrays
+// in a zippered iteration, they must match not only in terms of size, but
+// also shape.
```
Comment (vasslitvinov) [inline test/release/examples/primers/iterators.chpl:71]: I suggest explaining what "shape" is, ex.: `... also shape, i.e., the size in each dimension.`

**Review by vasslitvinov — APPROVED** [review]
Looks good. Please see my comment.

I did not see an indication that the design question in #20051 has been resolved. I am assuming that it has been.

The spec does not talk about zippered iteration with unbounded ranges, so it does not need to be updated. The primer that you updated should suffice.

File: compiler/resolution/lowerIterators.cpp (hunk 1 above)
Comment (bradcray) [reply compiler/resolution/lowerIterators.cpp:2531]: I'm not a huge fan of leaning too much on issue numbers in error messages (though I know we do it sometimes) for fear of someone not having internet access.  Good point about the lack of an explanation though.  I'm updating to:

> The behavior of zippered serial loops driven by unbounded ranges has been fixed in this release to act as though they were conceptually infinite; to maintain the previous behavior, swap a bounded iterand into the first expression of the 'zip(...)'.

if that sounds OK to you.

File: test/release/examples/primers/iterators.chpl (hunk 2 above)
Comment (bradcray) [reply test/release/examples/primers/iterators.chpl]: Done in https://github.com/chapel-lang/chapel/pull/20891/commits/368dc78d91ab808b363b6f10dcc8f7d163ea885a

**Comment by bradcray** [comment]
Thanks for reviewing, Vass.  W.r.t. your high-level comments:

> I did not see an indication that the design question in https://github.com/chapel-lang/chapel/issues/20051 has been resolved. I am assuming that it has been.

I advertised a design issue internally on Oct 21 which got a few thumbs-ups and no pushback, so I was planning on proceeding on that basis.  You're correct that I hadn't posted an intention to proceed with the plan on #20051 yet, mostly to maximize the chances for people to object before merging.  I've done that now.

> The spec does not talk about zippered iteration with unbounded ranges, so it does not need to be updated. 

Thanks for thinking of that.  I suspect that we should put more text into the spec about zippered iterators, and will plan to do that either here before merging or as a follow-up depending on how quickly you OK the revised error message.

**Comment by vasslitvinov** [comment]
I like the revised error message. It is a "go" for me.

Speaking of which, we need to ensure that the behavior in the case "about to overflow the index type of the unbounded range" is consistent with the non-zippered case. This could be a follow-up item.

For the OP/merge message, let us make it explicit that it will be an error in the zip(unbounded, bounded) case if the loop exhausts all indices in the second iterand without breaking/returning/halting.

**Comment by bradcray** [comment]
> we need to ensure that the behavior in the case "about to overflow the index type of the unbounded range" is consistent with the non-zippered case. 

To make sure I'm understanding: You're suggesting a test in which we make sure something like `zip(250:uint(8).., 1..)` works as expected?

**Comment by vasslitvinov** [comment]
Correct, adding tests is the way to go. Iterating over `zip(250:uint(8).., 1..)` and `zip(250:uint(8).., 1..10)` should report the same error as iterating over `250:uint(8)..`. So that would be two tests.

**Comment by bradcray** [comment]
@vasslitvinov : I've just pushed the new test (one test with multiple cases).  For a typical compilation, the `zip(250:int(8).., 1..10)` case actually reports a "zipped expressions have mismatched lengths" style error, and the reason is that the compiler injects the check for that problem just after the zippered loops while the halt due to running off of the unbounded is in the iterator itself and after the loop that the compiler is zipping.  But that seems reasonable to me since it is accurate.  When running with `--baseline` where the zippering is unoptimized, we get the other error.  I think that this is acceptable since both messages are accurate and don't think it's worth the effort to try and unify them at this point.  The test just needed some extra care to work in that case as well.  This is a good reminder to check the other tests I wrote for `--baseline`.

**Comment by bradcray** [comment]
Tests modified by this PR work under `--baseline`, so I think this is good to go once CI finishes.

**Pull request merged**